R/infrastructure-git.R

```diff
@@ -29,38 +29,90 @@ use_git <- function(message = "Initial commit", pkg = ".") {
+#'
+#'   The argument \code{protocol} reflects how you wish to authenticate with
+#'   GitHub for this repo in the long run. For either \code{protocol}, a remote
+#'   named "origin" is created, an initial push is made using \code{auth_token}
+#'   for authentication, and a remote tracking branch is set. The URL of the
+#'   "origin" remote has the form
+#'   \code{https://github.com/<USERNAME>/<REPO>.git} (\code{protocol = "https"})
+#'   or \code{git@@github.com:<USERNAME>/<REPO>.git} (\code{protocol = "ssh"},
+#'   the default). Read GitHub's help
+#'   \href{https://help.github.com/articles/which-remote-url-should-i-use/}{Which
+#'    remote URL should I use?} for more information.
 #'
 #' @inheritParams install_github
 #' @inheritParams use_git
 #' @param private If \code{TRUE}, creates a private repository.
+#' @param protocol transfer protocol, either "ssh" (the default) or "https"
 #' @family git infrastructure
 #' @keywords internal
 #' @export
-use_github <- function(auth_token = github_pat(), private = FALSE, pkg = ".") {
+#' @examples
+#' \dontrun{
+#' create("testpkg")
+#' use_github(pkg = "testpkg", protocol = "https", private = TRUE)
+#' }
+use_github <- function(auth_token = github_pat(), private = FALSE, pkg = ".",
+                       protocol = c("ssh", "https")) {
   if (is.null(auth_token)) {
     stop("GITHUB_PAT required to create new repo")
   }
 
+  protocol <- match.arg(protocol)
+
   pkg <- as.package(pkg)
   use_git(pkg = pkg)
 
-  r <- git2r::repository(pkg$path)
-  if ("origin" %in% git2r::remotes(r))
+  if (uses_github(pkg$path))
     return(invisible())
 
-  message("Creating GitHub repository")
-
+  message("* Creating GitHub repository")
   create <- github_POST("user/repos", pat = auth_token, body = list(
     name = jsonlite::unbox(pkg$package),
-    description = jsonlite::unbox(pkg$description),
+    #description = jsonlite::unbox(gsub("\n", " ", pkg$description)),
+    description = jsonlite::unbox(pkg$title),
     private = jsonlite::unbox(private)
   ))
 
-  message("Adding remote to GitHub")
-  git2r::remote_add(r, "origin", create$ssh_url)
-  # git2r::branch_set_upstream(git2r::head(r), "origin/master")
-  # git2r::push(r, "origin", "refs/heads/master")
+  message("* Adding GitHub remote")
+  r <- git2r::repository(pkg$path)
+  ## use https universally at this point! we deal with ssh case below
+  git2r::remote_add(r, "origin", create$clone_url)
+
+  message("* Adding GitHub links to DESCRIPTION")
+  use_github_links(pkg$path)
+  if (git_uncommitted(pkg$path)) {
+    git2r::add(r, "DESCRIPTION")
+    git2r::commit(r, "Add GitHub links to DESCRIPTION")
+  }
+
+  message("* Pushing to GitHub and setting remote tracking branch")
+  ## in https case, when GITHUB_PAT is passed as password,
+  ## the username is immaterial, but git2r doesn't know that
+  cred <- git2r::cred_env("DUMMY_EMAIL", "PAT")
+  with_envvar(c("DUMMY_EMAIL" = "whatever", "PAT" = auth_token),
```

Comment:
I just learned, with some pain, that if you send an empty email, `git2r::push()` hangs the R process (and RStudio somehow can't recover). I'll bring that up over in `git2r`. But for now I'm going with `EMAIL = "EMAIL"` here. Once `git2r` updates, we can use the new `git2r::cred_token()` and will no longer need to provide dummy email.